### PR TITLE
mem-ruby: Added WIB State to VIPER TCC Cache

### PR DESCRIPTION
**Pull request by DanKouch**
Added WIB (Waiting on Writethrough Ack; Will be Bypassed) state which is transitioned to when a dirty line in the TCC is evicted in a bypassed read. Previously, we were transitioning to invalid.

While a WI (Waiting on Writethrough Ack) state exists, transitions from it on WBAck deallocates the TBE, which contains SLC bit information needed to trigger the Bypass event when the read response from the directory comes in.

Without this change, WB acknowledgements from the directory in read bypass evicts (with the SLC bit set) were being treated as if they were read responses, leading to an invalid transition panic.

Change-Id: I703c3fe8af0366856552bb677810cb1a8f2896de

### CODE REVIEW

File: src/mem/ruby/protocol/GPU_VIPER-TCC.sm
Comment (mattsinc) [inline src/mem/ruby/protocol/GPU_VIPER-TCC.sm]: minor nit: "desc" should start 1 character further to the left

File: src/mem/ruby/protocol/GPU_VIPER-TCC.sm
Comment (mattsinc) [inline src/mem/ruby/protocol/GPU_VIPER-TCC.sm:298]: Please add a comment here explaining what this is for

**Review by mattsinc — CHANGES_REQUESTED** [review]

**Comment by DanKouch** [comment]
@mattsinc Since I don't believe GitHub has the same concept of patch sets as Gerrit does, do you know if minor tweaks should be made as additional commits, or by updating the original commit and force-pushing it? The later seems more similar to the Gerrit method, but I haven't seen it used often.

Also, I see that the CI check for change IDs only requires one commit in a PR to have a change ID. Do you happen to know if minor commits should have one?

**Comment by mattsinc** [comment]
> @mattsinc Since I don't believe GitHub has the same concept of patch sets as Gerrit does, do you know if minor tweaks should be made as additional commits, or by updating the original commit and force-pushing it? The later seems more similar to the Gerrit method, but I haven't seen it used often.
> 
> Also, I see that the CI check for change IDs only requires one commit in a PR to have a change ID. Do you happen to know if minor commits should have one?

From the email from @BobbyRBruce : "Note that on GitHub, we will be following a squash+merge contribution model. Therefore, pull requests (PRs) on GitHub will be a bit different from relation chains on Gerrit. We ask that when you create PRs on GitHub that each PR is for one logical change (unlike Gerrit relation chains that have often included many different logical changes). We also ask that each PR is as small and self-contained as possible."

So I'm pretty sure that means force-pushing.

**Comment by BobbyRBruce** [comment]
> > @mattsinc Since I don't believe GitHub has the same concept of patch sets as Gerrit does, do you know if minor tweaks should be made as additional commits, or by updating the original commit and force-pushing it? The later seems more similar to the Gerrit method, but I haven't seen it used often.
> > Also, I see that the CI check for change IDs only requires one commit in a PR to have a change ID. Do you happen to know if minor commits should have one?
> 
> From the email from @BobbyRBruce : "Note that on GitHub, we will be following a squash+merge contribution model. Therefore, pull requests (PRs) on GitHub will be a bit different from relation chains on Gerrit. We ask that when you create PRs on GitHub that each PR is for one logical change (unlike Gerrit relation chains that have often included many different logical changes). We also ask that each PR is as small and self-contained as possible."
> 
> So I'm pretty sure that means force-pushing.

For now, both are acceptable, but I think adding commits to the end of the pull-request gives a nice history of the pull-request over time. I'm not sure what's maintained if you force push.

File: src/mem/ruby/protocol/GPU_VIPER-TCC.sm
Comment (DanKouch) [reply src/mem/ruby/protocol/GPU_VIPER-TCC.sm]: Fixed.

File: src/mem/ruby/protocol/GPU_VIPER-TCC.sm
Comment (DanKouch) [reply src/mem/ruby/protocol/GPU_VIPER-TCC.sm:298]: Added.

**Comment by DanKouch** [comment]
@mattsinc @BobbyRBruce I tried force-pushing to see what would happen. It looks like GitHub is able to tell what's going on and allows you to compare between the force-pushed commits within the PR.

**Review by mattsinc — DISMISSED** [review]
Ideally Matt P should review this too, but it doesn't seem we're able to add him on Git yet ...

**Review by powerjg — DISMISSED** [review]
lgtm. @mattsinc, you can hit the "squash and merge" button if you want :).

If you want Matt P. to take a look, reach out to him and ask him to accept the invite that we sent.

**Review by abmerop — COMMENTED** [review]
I think this should also handle L2_Repl and PrbInv events in the same way as transition *from* WI, otherwise looks good.

**Review by abmerop — CHANGES_REQUESTED** [review]
Sorry - I suppose I needed to use the request changes option.

Please add transitions for L2_Repl and PrbInv events when in WIB state.  They should be the same as the transitions from WI state.

**Comment by BobbyRBruce** [comment]
> @mattsinc @BobbyRBruce I tried force-pushing to see what would happen. It looks like GitHub is able to tell what's going on and allows you to compare between the force-pushed commits within the PR.

Thanks! That's really good to know.

**Review by BobbyRBruce — CHANGES_REQUESTED** [review]
From Matt P.

>Please add transitions for L2_Repl and PrbInv events when in WIB state. They should be the same as the transitions from WI state.

**Review by BobbyRBruce — DISMISSED** [review]
It appears to me all comments have been addressed.

**Comment by mattsinc** [comment]
> Sorry - I suppose I needed to use the request changes option.
> 
> Please add transitions for L2_Repl and PrbInv events when in WIB state. They should be the same as the transitions from WI state.

@abmerop for these, wouldn't we want to stall them?  We are waiting for a reply to bypass, but these events are immediately evicting.  When the ACK comes back after we evicted it, I guess we just accept it came for something in this situation?

**Comment by abmerop** [comment]
> > Sorry - I suppose I needed to use the request changes option.
> > Please add transitions for L2_Repl and PrbInv events when in WIB state. They should be the same as the transitions from WI state.
> 
> @abmerop for these, wouldn't we want to stall them? We are waiting for a reply to bypass, but these events are immediately evicting. When the ACK comes back after we evicted it, I guess we just accept it came for something in this situation?

Stalling on an L2_Repl would be very dangerous and would probably deadlock.  Probe I am not sure, however when both of these are seen while in WI state, they remain in WI state and perform some actions. Essentially I am asking for the following:

```
transition({A, IV, WI, WIB}, L2_Repl) {
    i_invL2;
}

transition({A, IV, WI, WIB}, PrbInv) {
    pi_sendProbeResponseInv;
    pp_popProbeQueue;
}
```

This situation I think is currently very hard to test in gem5 but there could be a case where a replacement is need/probe comes in while in WIB state.

**Comment by mattsinc** [comment]
> > > Sorry - I suppose I needed to use the request changes option.
> > > Please add transitions for L2_Repl and PrbInv events when in WIB state. They should be the same as the transitions from WI state.
> > 
> > 
> > @abmerop for these, wouldn't we want to stall them? We are waiting for a reply to bypass, but these events are immediately evicting. When the ACK comes back after we evicted it, I guess we just accept it came for something in this situation?
> 
> Stalling on an L2_Repl would be very dangerous and would probably deadlock. Probe I am not sure, however when both of these are seen while in WI state, they remain in WI state and perform some actions. Essentially I am asking for the following:
> 
> ```
> transition({A, IV, WI, WIB}, L2_Repl) {
>     i_invL2;
> }
> 
> transition({A, IV, WI, WIB}, PrbInv) {
>     pi_sendProbeResponseInv;
>     pp_popProbeQueue;
> }
> ```
> 
> This situation I think is currently very hard to test in gem5 but there could be a case where a replacement is need/probe comes in while in WIB state.

I understand, but yeah I'm thinking of the case you're describing -- basically what happens when WIB is waiting on something but it gets evicted (by Probe/L2_Repl) while waiting on that ACK.  I guess as long as we don't have packets dropped it would be fine to just ignore an ACK if it arrives for a line that no longer exists after?  Except if there is another write after the Probe/Repl and this write also gets to WIB before the original ACK arrives -- admittedly this is a probably rare situation though.

Ultimately I'm fine letting this through and fixing the above if it happens later.  But it does seem plausible ...

**Review by mattsinc — DISMISSED** [review]
Carrying forward approval from @abmerop .

**Comment by mattsinc** [comment]
@BobbyRBruce how do we get more information on why this happened?

Test: pyunit-tests-ALL-x86_64-opt Failed

That's all the log says...

**Comment by powerjg** [comment]
> how do we get more information on why this happened?

You can download the zip file of the test that contains *everything* and dig in. It's available on this page: https://github.com/gem5/gem5/actions/runs/5535963692?pr=67

Here's the error. I'm not sure why this would fail. @BobbyRBruce??

```
======================================================================
ERROR: setUpClass (stdlib.resources.pyunit_workload_checks.WorkloadTestSuite)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "src/python/gem5/utils/filelock.py", line 64, in acquire
FileExistsError: [Errno 17] File exists: '/github/home/.cache/gem5/x86-ubuntu-18.04-img.lock.lock'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.10/unittest/mock.py", line 1369, in patched
    return func(*newargs, **newkeywargs)
  File "/__w/gem5/gem5/tests/pyunit/stdlib/resources/pyunit_workload_checks.py", line 138, in setUpClass
    cls.workload = Workload("simple-boot", gem5_version="develop")
  File "src/python/gem5/resources/workload.py", line 229, in __init__
  File "src/python/gem5/resources/resource.py", line 631, in obtain_resource
  File "src/python/gem5/resources/downloader.py", line 248, in get_resource
  File "src/python/gem5/utils/filelock.py", line 107, in __enter__
  File "src/python/gem5/utils/filelock.py", line 85, in acquire
gem5.utils.filelock.FileLockException: Timeout occured. This is likely due to the existence of the lock file '/github/home/.cache/gem5/x86-ubuntu-18.04-img.lock.lock'. If there's no other process the lock file, you can manually delete the lock file and rerun the script.
```

**Comment by mattsinc** [comment]
> > how do we get more information on why this happened?
> 
> You can download the zip file of the test that contains _everything_ and dig in. It's available on this page: https://github.com/gem5/gem5/actions/runs/5535963692?pr=67
> 
> Here's the error. I'm not sure why this would fail. @BobbyRBruce??
> 
> ```
> ======================================================================
> ERROR: setUpClass (stdlib.resources.pyunit_workload_checks.WorkloadTestSuite)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "src/python/gem5/utils/filelock.py", line 64, in acquire
> FileExistsError: [Errno 17] File exists: '/github/home/.cache/gem5/x86-ubuntu-18.04-img.lock.lock'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/usr/lib/python3.10/unittest/mock.py", line 1369, in patched
>     return func(*newargs, **newkeywargs)
>   File "/__w/gem5/gem5/tests/pyunit/stdlib/resources/pyunit_workload_checks.py", line 138, in setUpClass
>     cls.workload = Workload("simple-boot", gem5_version="develop")
>   File "src/python/gem5/resources/workload.py", line 229, in __init__
>   File "src/python/gem5/resources/resource.py", line 631, in obtain_resource
>   File "src/python/gem5/resources/downloader.py", line 248, in get_resource
>   File "src/python/gem5/utils/filelock.py", line 107, in __enter__
>   File "src/python/gem5/utils/filelock.py", line 85, in acquire
> gem5.utils.filelock.FileLockException: Timeout occured. This is likely due to the existence of the lock file '/github/home/.cache/gem5/x86-ubuntu-18.04-img.lock.lock'. If there's no other process the lock file, you can manually delete the lock file and rerun the script.
> ```

Thanks @powerjg .  @BobbyRBruce yes please let us know if you know anything about that failure.

**Comment by DanKouch** [comment]
Re-pushed and rebased to re-run CI testing.

**Review by mattsinc — APPROVED** [review]
Doesn't seem like anything changed?

**Comment by powerjg** [comment]
Is this ready to submit?

**Comment by abmerop** [comment]
Seems I am not authorized to merge.  I think there are still outstanding changes though.

**Comment by mattsinc** [comment]
> Seems I am not authorized to merge. I think there are still outstanding changes though.

What outstanding changes do you see?  I thought it was ready to go, but maybe I missed something ...

I am able to squash and merge (maybe you need to accept the Maintainer invite Matt?), but was going to let Daniel do this since it's his patch.

**Review by abmerop — APPROVED** [review]
Apologies, I see the changes now: LGTM.

I also have no idea where said maintainer invites are going but i've checked all 3 emails attached to this account and looked around github and saw nothing. Might need someone to explain it to me.

**Comment by mattsinc** [comment]
> Apologies, I see the changes now: LGTM.
> 
> I also have no idea where said maintainer invites are going but i've checked all 3 emails attached to this account and looked around github and saw nothing. Might need someone to explain it to me.

@mkjost0 see @abmerop above's message, about not being able to be added to maintainers

**Comment by abmerop** [comment]
Jason added me and I merged this one as a test

**Pull request merged**
